### ISSUE TITLE
kubernetes pod-to-pod low bandwidth 

### ISSUE THREAD

**Issue by ali-raza-tariq** [kubernetes/kubernetes#95860]
**What happened**: 
Hello everyone! i am kinda new to Kubernetes setup and don’t know a lot about its network settings. I did a test setup on 3 bare-metal nodes (running v1.19.3 using flannel cni). The nodes are connected via 10Gbits/s links. When i ran iperf on two different nodes, it showed a bandwidth of 9.4GBits/s which is what i would expect. But when i ran two pods on these two nodes and did iperf from within the pods, it could only achieve about somewhere between 1 to 4 GBits/s. Is this the expected behavior? or is there some network setting i need to configure?

**What you expected to happen**:
I would expect the perf to saturate the physical links capacity.

**How to reproduce it (as minimally and precisely as possible)**:
Simply start two idle containers on different nodes and perform iperf test.

**Anything else we need to know?**:
I have tested it on Kubernetes v1.14.3 as well with similar results.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.19.3
- Cloud provider or hardware configuration: N/A
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04
- Kernel (e.g. `uname -a`): N/A
- Install tools: iperf
- Network plugin and version (if this is a network-related bug): flannel CNI
- Others: N/A

**Comment by hanlins** [comment]
Hi @ali-raza-tariq, flannel is an overlay CNI solution and it will involve encapsulation/decapsulation. Have you tried other solution? If throughput is your concern maybe you can try Calico with pure L3 pod network, that eliminates the encapsulation overhead.

**Comment by ali-raza-tariq** [comment]
Hi @hanlins, thank you for the quick response. I did try Calico which improved the performance a little (was able to hit 6GBits/s) but was again unable to saturate the link. Also, while I was looking into this issue I found a related benchmark article that showed how these CNI are capable to saturating the 10GBit/s link (https://itnext.io/benchmark-results-of-kubernetes-network-plugins-cni-over-10gbit-s-network-36475925a560). Hence, it has become a bit confusing and I just want to make sure that my network configuration is not putting an artificial bottleneck!

**Comment by hanlins** [comment]
Hi @ali-raza-tariq , thanks for the feedback! From this [doc](https://github.com/coreos/flannel/blob/master/Documentation/running.md#running-manually) seems flannel's MTU is 1450 within container, can you do the same thing and report the MTU in your setup? Also please share the MTU of the interface on your bare metal machine, that would be helpful. Also you can try tune MTU according to [flannel doc](https://coreos.com/flannel/docs/latest/flannel-config.html).

**Comment by ali-raza-tariq** [comment]
I have tried changing the MTU for both physical and virtual interfaces with `ip link set` (again without any benefit). Please find attached below the network interfaces of master node & worker node. Let me know it you want me to set MTU to some specific value.
`master`
```
cni0      Link encap:Ethernet  HWaddr ee:c2:73:c4:f4:0f  
          inet addr:10.244.0.1  Bcast:10.244.0.255  Mask:255.255.255.0
          inet6 addr: fe80::ecc2:73ff:fec4:f40f/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:21966 errors:0 dropped:0 overruns:0 frame:0
          TX packets:30778 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:1224464381 (1.2 GB)  TX bytes:2758483 (2.7 MB)

docker0   Link encap:Ethernet  HWaddr 02:42:57:92:fb:d0  
          inet addr:172.17.0.1  Bcast:172.17.255.255  Mask:255.255.0.0
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eno1      Link encap:Ethernet  HWaddr ec:b1:d7:85:1a:32  
          inet addr:128.110.153.137  Bcast:128.110.155.255  Mask:255.255.252.0
          inet6 addr: fe80::eeb1:d7ff:fe85:1a32/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1376934 errors:0 dropped:0 overruns:0 frame:0
          TX packets:939345 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:1966127244 (1.9 GB)  TX bytes:1329320139 (1.3 GB)

eno1d1    Link encap:Ethernet  HWaddr ec:b1:d7:85:1a:33  
          inet addr:10.10.1.3  Bcast:10.10.1.255  Mask:255.255.255.0
          inet6 addr: fe80::eeb1:d7ff:fe85:1a33/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:92724 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2775358 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:6209133 (6.2 MB)  TX bytes:4199870478 (4.1 GB)

flannel.1 Link encap:Ethernet  HWaddr 6e:9d:65:4b:ef:ac  
          inet addr:10.244.0.0  Bcast:0.0.0.0  Mask:255.255.255.255
          inet6 addr: fe80::6c9d:65ff:fe4b:efac/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:28732 errors:0 dropped:0 overruns:0 frame:0
          TX packets:19979 errors:0 dropped:8 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:1510394 (1.5 MB)  TX bytes:1224326008 (1.2 GB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:120117 errors:0 dropped:0 overruns:0 frame:0
          TX packets:120117 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:30008779 (30.0 MB)  TX bytes:30008779 (30.0 MB)

veth4b48762c Link encap:Ethernet  HWaddr c2:49:4c:0f:3a:df  
          inet6 addr: fe80::c049:4cff:fe0f:3adf/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:20001 errors:0 dropped:0 overruns:0 frame:0
          TX packets:28758 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:1224606463 (1.2 GB)  TX bytes:1975829 (1.9 MB)

vethaf77077a Link encap:Ethernet  HWaddr 6a:ab:3d:47:17:18  
          inet6 addr: fe80::68ab:3dff:fe47:1718/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:992 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1045 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:84156 (84.1 KB)  TX bytes:394218 (394.2 KB)

vethd2b025f4 Link encap:Ethernet  HWaddr 72:43:9e:86:fe:5a  
          inet6 addr: fe80::7043:9eff:fe86:fe5a/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:987 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1028 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:83240 (83.2 KB)  TX bytes:392940 (392.9 KB)
```
`worker node`
```
cni0      Link encap:Ethernet  HWaddr 7a:a2:13:31:a0:ed  
          inet addr:10.244.1.1  Bcast:10.244.1.255  Mask:255.255.255.0
          inet6 addr: fe80::78a2:13ff:fe31:a0ed/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:28755 errors:0 dropped:0 overruns:0 frame:0
          TX packets:29329 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:1511697 (1.5 MB)  TX bytes:1225283691 (1.2 GB)

docker0   Link encap:Ethernet  HWaddr 02:42:97:06:cd:54  
          inet addr:172.17.0.1  Bcast:172.17.255.255  Mask:255.255.0.0
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eno1      Link encap:Ethernet  HWaddr 14:58:d0:58:9f:c2  
          inet addr:128.110.153.151  Bcast:128.110.155.255  Mask:255.255.252.0
          inet6 addr: fe80::1658:d0ff:fe58:9fc2/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:2100150 errors:0 dropped:0 overruns:0 frame:0
          TX packets:85505 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:3109148136 (3.1 GB)  TX bytes:7222415 (7.2 MB)

eno1d1    Link encap:Ethernet  HWaddr 14:58:d0:58:9f:c3  
          inet addr:10.10.1.4  Bcast:10.10.1.255  Mask:255.255.255.0
          inet6 addr: fe80::1658:d0ff:fe58:9fc3/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:2775263 errors:0 dropped:0 overruns:0 frame:0
          TX packets:92921 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:4199445576 (4.1 GB)  TX bytes:6240757 (6.2 MB)

flannel.1 Link encap:Ethernet  HWaddr 36:18:30:a8:14:06  
          inet addr:10.244.1.0  Bcast:0.0.0.0  Mask:255.255.255.255
          inet6 addr: fe80::3418:30ff:fea8:1406/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:875036 errors:0 dropped:0 overruns:0 frame:0
          TX packets:28732 errors:0 dropped:8 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:1268788972 (1.2 GB)  TX bytes:1510394 (1.5 MB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:922 errors:0 dropped:0 overruns:0 frame:0
          TX packets:922 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:56930 (56.9 KB)  TX bytes:56930 (56.9 KB)

veth2d7bf0e6 Link encap:Ethernet  HWaddr b2:a4:e3:4e:33:f1  
          inet6 addr: fe80::b0a4:e3ff:fe4e:33f1/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1450  Metric:1
          RX packets:28755 errors:0 dropped:0 overruns:0 frame:0
          TX packets:29335 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:1914267 (1.9 MB)  TX bytes:1225284123 (1.2 GB)
```

**Comment by hanlins** [comment]
Hi @ali-raza-tariq I found [this doc](https://elkano.org/blog/testing-10g-network-iperf/) useful, in that setup the author increased the MTU to something way larger than default 1500 and saturated the network. I was trying to check if there was a difference between MTU values for your physical interface and the ones used in containers. If your original physical interface MTU is something larger than 1500 you can revert it and try change the veth to something larger (or just change flannel config and restart pods) and maybe also increase docker bridge and cni0's MTU accordingly. If the original physical interface MTU is 1450 then I think the issue might not be MTU related, but you can still try increase the MTU and see if it helps.

**Comment by ali-raza-tariq** [comment]
The original physical interfaces have an MTU of 1500. All other interfaces (veth, cni0, flannel) had a default value of 1450 except docker0 which also had a default value of 1500 MTU. I have tried out various combinations including what you suggested (increased all virtual interfaces MTU to 9000) again without any performance difference. For some reason, I couldn't increase the MTU for docker0 beyond 1500.

**Comment by hanlins** [comment]
I see. Btw have you tried deploying multiple pods simutaneouslly and check the sum of the total bandwidth? If your pods are sharing the bandwidth and reached 10G in total then it should be fine right? If you only want to achieve maximum bandwidth for a dedicated pod then you might simply set `hostNetwork` to true in your pod spec so it will directly use your node's network.

**Comment by ali-raza-tariq** [comment]
interesting! ... I did not think of that previously. This seems to be a pod-to-pod bandwidth limit. Having two different pairs of pods sharing the same physical link, I get 3.5GBits/s each. 

 ```If your pods are sharing the bandwidth and reached 10G in total then it should be fine right?```

yes it would make sense but the question remains why? I am working with distributed ML jobs and for some reason, the network was becoming the bottleneck. That is what lead me to this issue!

```for a dedicated pod then you might simply set hostNetwork to true in your pod spec so it will directly use your node's network```

that might not be possible since, one can have more than one container running in parallel. Not all the containers are going to use network at the same time so, multiplexing would make more sense for high-performance and efficiency.

So ideally, I would like to make sure that my setup is using resources to the potential capacity. If 4GBits/s is the best you can do with pod-to-pod using flannel ... then that's fine but I need to make sure that is indeed the case and not some network mis-configuration I didn't know about.

**Comment by hanlins** [comment]
Routing to existing issue: https://github.com/kubernetes/kubernetes/pull/90839 for more context
@jayunit100

**Comment by hanlins** [comment]
Hi @ali-raza-tariq, I think this is caused by the nature of container network. There's a [paper](http://ranger.uta.edu/~jrao/papers/INFOCOM18.pdf) on container network performance, and generally container network performance drops a lot, quote from the study:
> For example, Docker native overlay inflicted 82.8% throughput drop and 55% latency increase compared to the baseline.

I think that data matches your observation pretty well despite setup/version difference. In that study however, it didn't cover SR-IOV. I would say if your physical NIC supports SR-IOV and you're still interested in this issue, maybe you can try it. Here's a [blog](https://software.intel.com/content/www/us/en/develop/articles/single-root-inputoutput-virtualization-sr-iov-with-linux-containers.html#_Toc434917180) from intel using LXC + SR-IOV combination for your reference, basically in their setup, they achieved the same performance in container network as VM network.

**Comment by ali-raza-tariq** [comment]
@hanlins I understand there is an overhead which would be very clear in terms of latency. but my concern is regarding utilization of network capacity which should still be possible over longer periods, like I mentioned a benchmark in one of my comments, the article shows how you can still saturate the 10 Gbits/s link. Furthermore, I have been testing with various network configurations (my cluster had two IP address subnet - internal and external), forcing the internal network addresses I did once end-up with flannel CNI setup getting 9+ Gbits/s (consistently - and not just for a few seconds/minutes) so I can safely say its not because of the overlay overhead. Since I can not reproduce it always (I have been able to do multiple different setups same as above giving 9+ Gbits/s) I can not be sure of the cause.

**Comment by dcbw** [comment]
@ali-raza-tariq with flannel (which you reference above) remember that those "flannel.1" interfaces are VXLAN tunnels to other nodes, which requires encapsulation on the sender, and decapsulation on the receiver. That's why the MTU is lower, because the extra 50 bytes is required for the VXLAN header.

VXLAN requires some amount of processing for the encap/decap, that is one reason you get lower bandwidth. Many network cards include VXLAN offload capability, which accelerates these operations. You can check if your NIC supports it by running `ethtool -k <nic> | grep tx-udp` and see if any of those are "on". If they are off, and you do *not* see [fixed] (which indicates the NIC does not support offload of that feature) you can try to turn them on.

A NIC with VXLAN offload typically allows 90%+ of native performance.

**Comment by dcbw** [comment]
In any case, this is not a Kubernetes issue, it's specific to the network plugin in-use. If you'd like to follow up, please do so with the flannel project (or other network plugin). Thanks!

**Comment by ali-raza-tariq** [comment]
For others who might have the same problem, the issue was indeed with the network plugin (I was using flannel). My machines had multiple network interfaces (both 1GBits/s & 10GBits/s) so even though I configured Kubelet to use the `--node-ip` flag, the flannel would still use 1GBits/s for some servers. Since the interface names were different on different worker nodes, I used `--iface-regex` to configure flannel. (- --iface-regex=10\.10\.*\.*, because the interface ips were of the form 10.10.x.x). Using this, it would consistently use the interface matching the regex. 

Thanks for all the help!
